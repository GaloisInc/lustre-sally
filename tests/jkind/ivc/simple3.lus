-- IVC for sub2~0.sub1~0.ok: sub2.sub1.b

node sub1(x : int) returns (y : int; assum : bool);
var
  a, b, ok : bool;
let
  y = x + (0 -> pre y);

  a = x >= -1;
  b = x >= 0;
  assum = a and b;

  ok = y >= 0;
  --%IVC a, b;
  --%PROPERTY ok;
tel;

node sub2(x : int) returns (y : int; assum : bool);
let
  y, assum = sub1(x);
tel;

node main(x : int) returns (y : int);
var
  assum : bool;
let
  y, assum = sub2(x);
  assert assum;
tel;
