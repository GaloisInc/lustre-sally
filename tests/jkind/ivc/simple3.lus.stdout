Sally Model
===========

(define-state-type
   S
   (
     (a Bool)
     (assum Bool)
     (b Bool)
     (ok Bool)
     (x Int)
     (y Int)
     (|gal-initializing| Bool)
     (|init_pre_y:init| Bool)
     (|init_pre_y| Int)
     (|pre_y| Int)
     (|x:1| Int)
     (|x:2| Int))
   (
     (|x:2| Int)))
(define-transition-system
   TS
   S
   (and
      (= |gal-initializing| true)
      (= |init_pre_y:init| true))
   (and
      (= next.|gal-initializing| false)
      (= next.|x:2| input.|x:2|)
      next.assum
      (= next.|x:1| next.|x:2|)
      (= next.x next.|x:1|)
      (= next.|pre_y| state.y)
      (ite
         state.|init_pre_y:init|
         (= next.|init_pre_y| 0)
         (= next.|init_pre_y| next.|pre_y|))
      (= next.|init_pre_y:init| false)
      (=
         next.y
         (+ next.x next.|init_pre_y|))
      (=
         next.a
         (<=
            (- 1)
            next.x))
      (=
         next.b
         (<= 0 next.x))
      (=
         next.assum
         (and next.a next.b))
      (=
         next.ok
         (<= 0 next.y))))

(query
   TS
   (or |gal-initializing| ok))

[Lustre] Validating properties:
[Lustre] Property ok...
  [Sally][Valid] using 1 past state
[Lustre] Summary:
  [Valid] 1
  [Unknown] 0
  [Invalid] 0
[Lustre] Model status: [Valid]
